spanner/cloud-client/backup_sample.py

```diff
@@ -0,0 +1,273 @@
+
+
+# [START spanner_cancel_backup]
+def cancel_backup(instance_id, database_id, backup_id):
+    spanner_client = spanner.Client()
+    instance = spanner_client.instance(instance_id)
+    database = instance.database(database_id)
+
+    expire_time = datetime.utcnow() + timedelta(days=30)
+
+    # Create a backup.
+    backup = instance.backup(
+        backup_id, database=database, expire_time=expire_time)
+    operation = backup.create()
+
+    # Cancel backup creation.
+    operation.cancel()
+
+    # Cancel operations are best effort so either it will complete or
+    # be cancelled.
+    while not operation.done():
+        time.sleep(300)  # 5 mins
+
+    # Deal with resource if the operation succeeded.
+    if backup.exists():
+        print("Backup was created before the cancel completed.")
+        backup.delete()
+        print("Backup deleted.")
+    else:
+        print("Backup creation was successfully cancelled.")
+# [END spanner_cancel_backup]
+
+
+# [START spanner_list_backup_operations]
+def list_backup_operations(instance_id, database_id):
+    spanner_client = spanner.Client()
+    instance = spanner_client.instance(instance_id)
+
+    # List the CreateBackup operations.
+    filter_ = (
+        "(metadata.database:{}) AND "
+        "(metadata.@type:type.googleapis.com/"
+        "google.spanner.admin.database.v1.CreateBackupMetadata)"
+    ).format(database_id)
+    operations = instance.list_backup_operations(filter_=filter_)
+    for op in operations:
+        metadata = op.metadata
+        # List the pending backups on the instance.
+        if metadata.progress.progress_percent < 100:
+            print("Backup {} on database {} pending: {}% complete.".format(
+                metadata.name, metadata.database,
+                metadata.progress.progress_percent))
+# [END spanner_list_backup_operations]
+
+
+# [START spanner_list_database_operations]
+def list_database_operations(instance_id):
+    spanner_client = spanner.Client()
+    instance = spanner_client.instance(instance_id)
+
+    # List the progress of restore.
+    filter_ = (
+        "(metadata.@type:type.googleapis.com/"
+        "google.spanner.admin.database.v1.OptimizeRestoredDatabaseMetadata)"
+    )
+    operations = instance.list_database_operations(filter_=filter_)
+    for op in operations:
+        print("Database {} restored from backup is {}% optimized.".format(
+            op.metadata.name, op.metadata.progress.progress_percent))
+# [END spanner_list_database_operations]
+
+
+# [START spanner_list_backups]
+def list_backups(instance_id):
+    spanner_client = spanner.Client()
+    instance = spanner_client.instance(instance_id)
+
+    # List all backups.
+    print("All backups:")
+    for backup in instance.list_backups():
+        print(backup.name)
+
+    # List all backups that contain a name.
+    print("All backups with backup name containing \"users\":")
+    for backup in instance.list_backups(filter_="name:users"):
```

Comment:
I got a question that does this `list_backups` return any results? I wonder whether or not there is a backup with `name:users`.

Reply:
Yeah it probably wouldn't because we can't control what people would put in the command line. For Node.js, I'm filtering on the backup ID passed into the function but seems like Python doesn't have that.

Reply:
Ok so Node.js, PHP and Go are more or less consistent. I've asked Knut to do the same in Java. The changes we ended up making are:

1. Get the backup ID as input to listBackups and filter on name:backupId.
2. Database filter was changed to filter on the databaseId given as input.
3. Expire time filter was changed to expire_time < (time now + 30 days).
4. size_bytes > X where X depends on the size of the database you're creating in other parts of the test. PHP is using X = 500 and Node.js is using X = 100.
5. create_time >= (time now - 1 day)

Hope it's ok to make these changes.